features/253-auth-sdk-callback-in-next/api.md

```diff
@@ -0,0 +1,209 @@
+| username                 |
+| email                    |
+| phone                    |
+| .....                    |
+|                          |
++--------------------------+
+
+// user record
+skygear.auth.currentUser;
+
+// current user auth info
+skygear.auth.isCurrentUserVerified;
+skygear.auth.isCurrentUserDisabled;
+
+// current user profile record
+const phone = skygear.auth.currentUser['phone'];
+
+// update record
+skygear.auth.currentUser['username'] = 'new-username';
+skygear.auth.currentUser['phone'] = 'new-phone';
+skygear.auth.updateUser(skygear.auth.currentUser);
+```
+
+# Changes on SDK
+
+Skygear next server will help generate `record` data type for SDK, therefore SDKs are expected no major changes.
+
+Such changes are occurred in server side rather than SDK because:
+
+- SDK doesn't know record gear has zero knowledge about whether record configured or not.
+- Server side implementation would apply to multiple SDKs at once.
+- Server side implementation eliminates possible errors.
+
+SDK auth container will have following changes to fetch auth info of a user:
+
+- properties:
+    - isCurrentUserVerified;
+    - isCurrentUserDisabled;
+- APIs:
+    - Promise<Boolean> isUserVerified(user<userRecord>);
+    - Promise<Boolean> isUserDisabled(user<userRecord>);
+
+SDK auth container will be added an API to help update user auth info:
+
+- Promise<record> updateUser(user<record>);
+
+# Changes on auth gear
+
+Auth gear will keep current implementation, embed user profile record in the response:
+
+Old:
+
+```
+type AuthResponse struct {
+	UserID      string              `json:"user_id,omitempty"`
+	Profile     *skyconv.JSONRecord `json:"profile"`
+	Roles       []string            `json:"roles,omitempty"`
+	AccessToken string              `json:"access_token,omitempty"`
+	LastLoginAt *time.Time          `json:"last_login_at,omitempty"`
+	LastSeenAt  *time.Time          `json:"last_seen_at,omitempty"`
+}
+```
+
+New:
+
+```
+type AuthResponse struct {
+	UserID      string                 `json:"user_id,omitempty"`
+	Profile     map[string]interface{} `json:"profile"`
+	Roles       []string               `json:"roles,omitempty"`
+	AccessToken string                 `json:"access_token,omitempty"`
+	LastLoginAt *time.Time             `json:"last_login_at,omitempty"`
+	LastSeenAt  *time.Time             `json:"last_seen_at,omitempty"`
+}
+```
+
+And `AuthResponse.Profile` is a marshalled JSON object:
+
+```
+{
+    "_access": null,
+    "_created_at": "<datetime>",
+    "_created_by": "<user_id>",
+    "_id": "user\/<record_id>",
+    "_ownerID": "<user_id>",
+    "_recordID": "<record_id>",
+    "_recordType": "user",
+    "_type": "record",
+    "_updated_at": "<datetime>",
+    "_updated_by": "<user_id>",
+    "email": "abc@example.com"
+}
+```
+
+To handle the case of record gear missing, auth gear will add a new table to save user's profile.
+
+```
+CREATE TABLE _auth_user_profile (
+  user_id text REFERENCES _core_user(id),
```

Comment:
Maybe we should add unique key to user_id, every user should only have 1 auth_data

Reply:
Or we should call it profile for the field name? 🤔

Reply:
@carmenlau I added UNIQUE, and update the field name to `data`.